Feature/collide box

Sebastopaul opened the pull request:
Les collide boxes ont été ajoutées ainsi que les triggers, je suis en incapacité de tester suite à un souci technique, je te laisse gérer ça ou alors on réglera ça ensemble parce que c'est handicapant, en attendant je passe aux interfaces

Sebastopaul committed:
feat: Adding CollideBox implementation

Sebastopaul committed:
Merge branch 'develop' into feature/collideBox

Sebastopaul committed:
feat: Adding OnCollideCheck method to CollideBox

Sebastopaul committed:
fix: Added rigidbody handlingin rollback method of CollideBoxSystem

Sebastopaul committed:
fix: Trying to fix collisions when entering and to add collisons when exiting a collideBox

Sebastopaul committed:
fix: Fixing rollbacks

Sebastopaul committed:
fix: Fixing exit from collideBox

skrilax91 commented on include/KappaEngine/Components/CollideBox.hpp:
```diff
@@ -1,16 +1,35 @@
+            std::string _tag;
+            std::vector<std::string> _notCollidingTags;
+
+            std::function<bool(std::shared_ptr<KappaEngine::Entity>, std::shared_ptr<KappaEngine::Entity>)> _onCollideEnter;
```
ici implémente des fonctions de base qui return false

skrilax91 replied:
Et créer des setter

skrilax91 commented on include/KappaEngine/Components/CollideBox.hpp:
```diff
@@ -1,16 +1,35 @@
+    class CollideBox : public IComponent {
+        public:
+            CollideBox(sf::FloatRect &collideBox, std::string &tag,
+                std::function<bool(std::shared_ptr<KappaEngine::Entity>, std::shared_ptr<KappaEngine::Entity>)> onCollideEnter =  nullptr,
```
Ne met pas ça dans le constructeur c'est imbuvable sinon

skrilax91 commented on src/Systems/CollideBoxSystem.cpp:
```diff
@@ -2,11 +2,151 @@
+                if (!findCollide(collideBox->_collided, otherCollideBox)) {
+                    enterCollideBox(entity, otherEntity, transform, rigidBody, collideBox, otherCollideBox);
+                } else {
+                    if (collideBox->_onCollideStay != nullptr)
```
Avec les modif au dessus plus besoin de ce check

skrilax91 commented on src/Systems/CollideBoxSystem.cpp:
```diff
@@ -2,11 +2,151 @@
+                                        Component::Transform *transform, Component::RigidBody *rigidBody,
+                                        Component::CollideBox *collideBox, Component::CollideBox *otherCollideBox)
+    {
+        if (collideBox->_onCollideExit)
```
Avec les modif au dessus plus besoin de ce check

skrilax91 commented on src/Systems/CollideBoxSystem.cpp:
```diff
@@ -2,11 +2,151 @@
+            else if (rigidBody)
+                rollbackOnExit(transform, rigidBody, collideBox, otherCollideBox);
+        else
+            collideBox->_collided.remove(*otherCollideBox);
```
Merge les deux remove en un seul, ça rendra le code plus propre.

skrilax91 commented on src/Systems/CollideBoxSystem.cpp:
```diff
@@ -2,11 +2,151 @@
+            collideBox->_collided.remove(*otherCollideBox);
+    }
+
+    void CollideBoxSystem::rollbackOnEnter(Component::Transform *transform, Component::RigidBody *rigidBody,
```
Attention, il faut que tu prenne en compte la position du transform, rappelle toi que la position de ton component est relatif à la position du transfor 


HINT: t'as juste a créer un nouveau floatRect qui combine les positions du transform et celle du component

skrilax91 replied:
Aussi met en paramètre de ta fonction uniquement ce dont tu as besoin, pas besoin d'appeler tout le component si tu n'utilise que une seule variable de celui ci

skrilax91 replied:
Il faut refaire la fonction plus simplement, voir les screen discord

skrilax91 commented on src/Systems/CollideBoxSystem.cpp:
```diff
@@ -2,11 +2,151 @@
+        }
+    }
+
+    void CollideBoxSystem::rollbackOnExit(Component::Transform *transform, Component::RigidBody *rigidBody,
```
Celle la aussi peut etre factorisé

skrilax91 commented on src/Systems/TriggerBoxSystem.cpp:
```diff
@@ -7,30 +7,27 @@
+            else if (!otherTrigger->enabled)
                 continue;
 
             if (trigger->_triggerBox.intersects(otherTrigger->_triggerBox)) {
```
Pareil ici, il faut créer un floatRect temporaire pour inclure la position du transform

skrilax91 replied:
Il faut également set les fonction OnTriggerEnter etc par defaut

Sebastopaul committed:
fix: Fixing rollback methods and rects handling

skrilax91 reviewed (dismissed):
A tester avant de merge  !

Sebastopaul committed:
fix: Fixing build errors

Sebastopaul committed:
Merge remote-tracking branch 'origin/develop' into feature/collideBox

Sebastopaul committed:
fix: Adding TriggerBox and CollideBox systems to the Manager loops, Fixing boxes not working at all.

The boxes were not taking scaling into account. Next step is fixing the rollback functions.

Sebastopaul committed:
Auto stash before merge of "feature/collideBox" and "origin/feature/collideBox"

Sebastopaul committed:
feat: Adding Networking to TriggerBox System

Sebastopaul committed:
feat: Adding Networking on collideBox

skrilax91 committed:
Merge branch 'develop' into feature/collideBox
